### ISSUE TITLE
Non-portable use of SAME_TYPE_AS

### ISSUE THREAD

**Issue by nncarlson** [nncarlson/petaca#4]
The source file [parameter_list_json.F90](https://github.com/nncarlson/petaca/blob/master/src/parameter_list_json.F90) makes use of "processor dependent" behavior of the SAME_TYPE_AS intrinsic function.  Specifically it applies the function to CLASS(*) variables A and B whose dynamic types are intrinsic types (which are _non-extensible_).  The Fortran 2008 standard says of the function, "If neither A nor B has extensible dynamic type, the result is processor dependent."  This needs to be replaced with a portable alternative.

**Comment by zmiimz** [comment]
mb something like this?

```fortran
   function  is_intrinsic_type(pp) result(answer)
      implicit none
      class(*), pointer, intent(in) :: pp
      logical :: answer
      select type (pp)
       type is (integer)
         answer = .true.
       type is (real)
         answer = .true.
       type is (complex)
         answer = .true.
       type is (character(len=*))
         answer = .true.
       type  is (logical)
         answer = .true.
       class default
         answer = .false.
      end select
   end function is_intrinsic_type

**Comment by zmiimz** [comment]
Ups. It seems that we need to specify intrinsic type of each possible kind in the "type is" guard... I am not sure now that this could be made  portable enough...

**Comment by nncarlson** [comment]
You are very close @zmiimz.  This would get awfully messy with all the possible kinds, except that in this particular case the dynamic type is limited to integer or real of one specific kind, or default logical or character.  I've worked up a replacement for SAME_TYPE_AS that has an internal function just like yours, except it returns a different integer depending on type.  The replacement function calls this for each of the arguments A and B and compares the integer "type" values.  I was going to commit this, until I realized I have no unit test for `parameter_list_json.F90` that would exercise the change, so I still have some testing to do.

I haven't yet run into any compilers for which SAME_TYPE_AS didn't do what I expected, though I've only tested Intel, NAG, and gfortran.  I'm trying now to get things to work with the Cray compiler and was thinking for time that it had different behavior here, but it turns out that its implementation of SAME_TYPE_AS is completely broken for CLASS(*) arguments of any type (compile time error).
